zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java

```diff
@@ -614,6 +619,51 @@ public Response updateParagraphConfig(@PathParam("noteId") String noteId,
     return new JsonResponse<>(Status.OK, "", p).build();
   }
 
+  /**
+   * Download paragraph
+   * @param noteId
+   * @param paragraphId
+   * @return data stream
+   * @throws IOException
+   */
+  @GET
+  @Produces("text/csv")
+  @Path("{notebookId}/paragraph/{paragraphId}/download")
+  @ZeppelinApi
+  public Response downParagraph(@PathParam("notebookId") String noteId,
+                                @PathParam("paragraphId") String paragraphId)
+          throws IOException {
+    LOGGER.info("Download paragraph, node.id: {}, paragrap.id: {}", noteId, paragraphId);
+
+    Note note = notebook.getNote(noteId);
+    if (note == null) {
+      return new JsonResponse(Status.NOT_FOUND, "note not found.").build();
+    }
+
+    Paragraph p = note.getParagraph(paragraphId);
+    if (p == null) {
+      return new JsonResponse(Status.NOT_FOUND, "paragraph not found.").build();
+    }
+
+    String resultFileParent = conf.getString(ZeppelinConfiguration.ConfVars.ZEPPELIN_RESULT_DATA_DIR);
+
+    String fileName = noteId + "_" + paragraphId;
+    final File file = new File(resultFileParent + "/" + fileName);
```

Comment:
Please use the following constructor, which takes the OS path separator into account. https://docs.oracle.com/javase/7/docs/api/java/io/File.html#File(java.lang.String,%20java.lang.String)

Reply:
thank you for review my code.
changed to OS path separator.
and replaced "/tmp" to System.getProperty("java.io.tmpdir")